### ISSUE TITLE
Querying M2A related Items

### ISSUE THREAD

**Issue by zanami** [directus/directus#4595]
Directus 9.0.0-rc51
    Postgres 12

I'm testing Directus 9 and very impressed so far except for one thing with m2a

TLDR> Looks like the "item" in a m2a junction table joins to the junction table itself.

My setup
Article collection with m2a relation called 'sections'
```
article
  id
  ...
  sections

articles_sections collection (auto-created)
  id
  articles_id
  item
  collection

hero collection (related)
  id (int)
  title
  ...

text collection (related)
  uuid (uuid)
  title
  ...
```

Notice the difference between hero and text collections, their PKs' types differ

Trying to query REST API

```
http://localhost:8055/items/articles/1?fields[]=*&fields[]=sections.item:*.*
{
  "data": {
    "id": 1,
    "label": "Article title",
    "picture": "2a228828-5fac-4063-93e7-db903203629b",
    "sections": [
        {
        "item": 1
        },
        {
        "item": 2
        },
        {
        "item": null
        }
    ]
  }
}
```
the third item above could not join articles_sections because the corresponding id is a string (uuid)

```
http://localhost:8055/items/articles/1?fields[]=*&fields[]=sections.*

{
  data: {
    id: 1,
    label: 'Article title',
    picture: '2a228828-5fac-4063-93e7-db903203629b',
    sections: [
      {
        id: 1,
        articles_id: 1,
        item: '1',
        collection: 'hero',
      },
      {
        id: 2,
        articles_id: 1,
        item: '2',
        collection: 'hero',
      },
      {
        id: 3,
        articles_id: 1,
        item: '14d7fc3f-156b-4ab7-b312-95956b6e6a73',
        collection: 'text',
      },
    ],
  },
};
```
From what I see, 'item' value should point to corresponding 'collection' table, not junction table.
Here's more

```
http://localhost:8055/items/articles/1?fields[]=*&fields[]=sections.*.*

{
  data: {
    id: 1,
    label: 'Article title',
    picture: '2a228828-5fac-4063-93e7-db903203629b',
    sections: [
      {
        id: 1,
        collection: 'hero',
        articles_id: {
          id: 1,
          label: 'Article title',
          picture: '2a228828-5fac-4063-93e7-db903203629b',
          sections: [1, 2, 3],
        },
        item: 1,
      },
      {
        id: 2,
        collection: 'hero',
        articles_id: {
          id: 1,
          label: 'Article title',
          picture: '2a228828-5fac-4063-93e7-db903203629b',
          sections: [1, 2, 3],
        },
        item: 2,
      },
      {
        id: 3,
        collection: 'text',
        articles_id: {
          id: 1,
          label: 'Article title',
          picture: '2a228828-5fac-4063-93e7-db903203629b',
          sections: [1, 2, 3],
        },
        item: null,
      },
    ],
  },
};

http://localhost:8055/items/articles_sections/3?fields[]=*

{
  "data": {
    "id": 3,
    "articles_id": 1,
    "item": "14d7fc3f-156b-4ab7-b312-95956b6e6a73",
    "collection": "text"
  }
}

http://localhost:8055/items/articles_sections/3?fields[]=*.*

{
  "data": {
    "id": 3,
    "collection": "text",
    "articles_id": {.......},
    "item": null
    }
}

```

Same happens in Directus if I try to set Display Template for the articles_sections collection (junction table) — endless recursion and no actual item's fields.

There is no such problem if you just get the raw 'item' and 'collection' values to query /{collection}/{item} which Directus UI actually does in most places.

**Comment by rijkvanzanten** [comment]
Hey @zanami! Would you be able to share a DB dump with this setup so I can easily reproduce this exact query locally on my end??

**Comment by zanami** [comment]
Sure! Attached below

ADMIN_EMAIL: 'local@test.com'
ADMIN_PASSWORD: '1234567'

This is a strange issue because I couldn't reproduce it from scratch using simplified setup (new set of containers, same config, similar data model, no updated updated_by fields etc.) — works fine.

Here's the new instance with similar setup

```
http://localhost:8055/items/article/1?fields[]=*&fields[]=sections.*.*

{
  data: {
    id: 1,
    heading: 'Test article',
    sections: [
      {
        id: 1,
        collection: 'hero',
        item: {
          id: 1,
          title: 'Hero Title',
        },
        article_id: {
          id: 1,
          heading: 'Test article',
          sections: [1, 2],
        },
      },
      {
        id: 2,
        collection: 'text',
        item: {
          uuid: '1fd43024-4257-4491-ab3a-84c479da77de',
          text: '<p>Custom text</p>',
        },
        article_id: {
          id: 1,
          heading: 'Test article',
          sections: [1, 2],
        },
      },
    ],
  },
};
```

Display Template behaves the same, though — endless recursion, no Item data

![Снимок экрана 2021-03-22 в 11 25 07](https://user-images.githubusercontent.com/4746069/111929208-6a533c00-8b01-11eb-9856-bf2b3077d544.png)

___

Then I added a collection (article) to the first ("broken") instance reusing existing hero/text collections as targets for a M2A field. Now 'hero' behaves the same (seems broken) and 'text' (with uuid) actually displays the item data.

```
http://localhost:8055/items/article/1?fields[]=*&fields[]=sections.*.*

{
  data: {
    id: 1,
    title: 'New test',
    sections: [
      {
        id: 1,
        collection: 'hero',
        article_id: {
          id: 1,
          title: 'New test',
          sections: [1, 2],
        },
        item: 2,
      },
      {
        id: 2,
        collection: 'text',
        article_id: {
          id: 1,
          title: 'New test',
          sections: [1, 2],
        },
        item: {
          uuid: '67cd0d44-b49a-49e2-a1b0-eefcb39182bc',
          text: '<p>another text</p>',
        },
      },
    ],
  },
};
```

Actually, article**s** collection now shows 'text' items correctly as well, so only 'hero' is broken.

With both 'article' and 'articles' collections (broken one)

` docker exec -u postgres -t database pg_dumpall -c -U directus > dump_`date +%d-%m-%Y"_"%H_%M_%S`.txt
`
[dump_22-03-2021_11_29_05.txt](https://github.com/directus/directus/files/6178885/dump_22-03-2021_11_29_05.txt)

Clean one

`docker exec -u postgres -t dbtest pg_dumpall -c -U directus > dump_`date +%d-%m-%Y"_"%H_%M_%S`.txt`

[dump_22-03-2021_11_31_29.txt](https://github.com/directus/directus/files/6178887/dump_22-03-2021_11_31_29.txt)

**Comment by zanami** [comment]
Also this happens with graphql
might be helpful

```
query {
	items {
		articles {
			id
		}
	}
}

Response

{
  "errors": [
    {
      "message": "Type hero must define one or more fields."
    }
  ]
}
```

![Снимок экрана 2021-03-22 в 16 58 58](https://user-images.githubusercontent.com/4746069/111952193-7e626200-8b30-11eb-9237-41163294e706.png)

Fields are there, obviously.

I ended up with removing and recreating the 'hero' collection, now it works. ¯\_(ツ)_/¯

**Comment by rijkvanzanten** [comment]
> I ended up with removing and recreating the 'hero' collection, now it works.

Tricky tricky!

**Comment by AlexisWalravens** [comment]
I have the same issue.(9.0.0-rc.53)
graphQL is no longer working since I created a collection that is used in a m2a relation.
```
{
  "errors": [
    {
      "message": "Type paragraph_text_image must define one or more fields."
    }
  ]
}
```

Have you identified what's causing it ?

**Comment by AlexisWalravens** [comment]
It as something to do with permissions.
When I turn off the read permission for the public role on paragraph_text_image it works again.
And now even if I turn it back on and off it works either way.

**Comment by rijkvanzanten** [comment]
@zanami I wasn't able to reproduce this on a clean instance (just like you mentioned it went away on a new one), however, I did notice a funky thing in your export!

After aligning the records in the dump, I noticed the following:

![CleanShot 2021-05-04 at 20 09 50@2x](https://user-images.githubusercontent.com/9141017/117084244-b6a2b600-ad14-11eb-9cae-d66159452e82.png)

(1) For whatever reason, the order is kinda strange (not important)
(2) The data for `item` under `articles_sections` seems to be.. shifted left a column?

I have no clue what might've caused that, but in case you run into it again, please let me know 😄
